### ISSUE TITLE
Why not a Better Ranking system for the URLs ?

### ISSUE THREAD

**Issue by lakshmanaram** [google/physical-web#608]
Hey,
Under the "Client" heading of the "documentation/technical_overview.md" file, it is stated that 'If you open the app, it lists the nearby beacons that it can see, sorted by signal strength'. So I assume that the ranking system currently takes only the signal strength into consideration when sorting the list of URLs. Considering the enormous scope of this project, there could easily be a situation wherein my browser detects 15+ URLs at some instant, for example in a mall. Then if I am a regular to a certain shop and i have clicked the shop's URL a number of times before in my previous visits, It would be nice if i could find this URL among the top few URLs in my list.

*Solution I am proposing :* Why can't we have a cache like structure locally in the mobile, which can store the URLs that we use frequently, and can later factor in while ranking the URLs along with the signal strength?

*PS: I have read through the documentation and many issues before posting this but was unable to find anything relevant to this topic. I am sorry if this has already been discussed about earlier.*

**Comment by scottjenson** [comment]
I completely agree. There are LOTS of things we'd like to do. We're just getting the ball rolling here. BTW, I should point out that how they are sorted is likely a very fertile area for exploration so there could be many types of solutions. It's one of the reasons we want this to be open. Opera may do something different from Google and so on. It's even possible that there could be version done for specific groups (e.g. a "foodie focused" browser).

But specifically to your question, to makes sense to use both broad anonymous usage data as well as personal data. The broad usage data would elevate the mall map URL as everyone that goes into the mal tends to click on it. Personal usage would reflect common shops you visit.

All of this should be on the table. In fact, there was even a discussion about support Progressive Web apps that you've saved to desktop. Those are clearly very high value use cases and should be ranked higher as well.

Bottom line: totally agree. We're working on improving our scanning to do stuff like this. We hope others are as well.

Scott

**Comment by lakshmanaram** [comment]
@scottjenson Thanks for the quick response. Actually I didn’t realize the bigger picture behind the sorting. As you said, It opens up a whole new level of personalization of the physical web as such. 

I am a student from India and I would like to contribute to this project by building a better ranking system using the personal usage data in the native app. I have read through the code style that is to be followed and the instructions to contribute. Can you assign me this task?

**Comment by scottjenson** [comment]
First, I'm thrilled you'd like to work on this. However, we are separate from the Google Summer of code and honestly, I'm not quite sure who to contact on that (I've trying to figure it out). In the mean time, please make sure you contact them directly.

As far as we are concerned, you can just contribute to us directly. As you are thinking of making your own scanner (which I agree is a very interesting idea!) Please just open your own repo and link to it from here whenever you discuss it. Once you get to a stable, working build, we'll be happy to put up a pointer on here to it.

**Comment by lakshmanaram** [comment]
@scottjenson I understand that contributing to this repo is completely different from Google summer of code contribution as they only want students to make services that use physical web. 

I thought when I wish to make any bug fix or develop a new feature, I am supposed to open up an issue and discuss about the solution and make sure that people aren’t already working on it before starting to work as given in the **CONTRIBUTING.md** file. I opened up this issue to get assigned to this task, so that i can start working on the changes I want to see. 
*Sorry if i had misunderstood anything.*

And I assumed that the native app client, found in this repo is just an example to show the working and is part of the pre-release of the actual physical web (in browsers). So If I work on this feature in the native app, and create a working build, Will my changes get merged to the master? Or I should only build my own new native from the scratch based on the app available in this repo to make all the changes I want to see.?

**Comment by scottjenson** [comment]
In general, it's safer for you to make your own branch and see how far you can go. I strongly encourage this as you don't answer to anyone and can be as creative an as experimental as you'd like.

Of course, if you strongly feel that you'd like to merge your changes into our source, you are certainly welcome to do that as well but that is a process that usually involves lots of discussion and some people find that a bit tedious at times.

Whatever path you choose, you are welcome and encouraged to discuss it here as an issue. The beauty of github is that both options are possible so there is little risk that you won't be able to explore your vision.

**Comment by lakshmanaram** [comment]
@scottjenson Cool, Thanks a lot for the immediate responses and guidance, I shall fork this repo and start working right away in a new branch. As you said I’ll think about wanting to merge & it’s necessity once I am done with the changes. :+1:
